tests/unit/websocket.cpp

```diff
@@ -0,0 +1,262 @@
+ public:
+  MockAuth() = default;
+  MockAuth(bool authentication, bool authorization, bool has_any_users)
+      : authentication_{authentication}, authorization_{authorization}, has_any_users_{has_any_users} {}
+
+  bool Authenticate(const std::string & /*username*/, const std::string & /*password*/) const override {
+    return authentication_;
+  }
+
+  bool HasUserPermission(const std::string & /*username*/, auth::Permission /*permission*/) const override {
+    return authorization_;
+  }
+
+  bool HasAnyUsers() const override { return has_any_users_; }
+
+ private:
+  bool authentication_{true};
+  bool authorization_{true};
+  bool has_any_users_{true};
+};
+
+class Client {
+ public:
+  Client() : ioc_{}, ws_{ioc_} {}
+
+  ~Client() { ws_.close(websocket::close_code::normal); }
+
+  void Connect(const std::string &host, const std::string &port) {
+    tcp::resolver resolver{ioc_};
+    auto endpoint_ = resolver.resolve(host, port);
+    auto ep = net::connect(ws_.next_layer(), endpoint_);
+    const auto server = host + ":" + std::to_string(ep.port());
+    ws_.set_option(websocket::stream_base::decorator([](websocket::request_type &req) {
+      req.set(http::field::user_agent, std::string(BOOST_BEAST_VERSION_STRING) + " websocket-client-coro");
+    }));
+
+    // Perform the websocket handshake
+    ws_.handshake(host, "/");
+  }
+
+  void Write(const std::string &msg) { ws_.write(net::buffer(msg)); }
+
+  std::string Read() {
+    ws_.read(buffer_);
+    const std::string response = beast::buffers_to_string(buffer_.data());
+    buffer_.consume(buffer_.size());
+    return response;
+  }
+
+ private:
+  net::io_context ioc_;
+  websocket::stream<tcp::socket> ws_;
+  beast::flat_buffer buffer_;
+};
+
+TEST(WebSocketServer, WebSockerServerCreation) {
+  /**
+   * Notice how there is not port management for the clients
+   * and servers, that is because when using "0.0.0.0" and address
+   * and 0 as port number we delegate port assignment to the os
+   * and he is the keeper of all available port numbers and
+   * assignees is automaticaly to us.
+   */
+  MockAuth auth{};
+  EXPECT_NO_THROW(communication::websocket::Server websocket_server({"0.0.0.0", 0}, &auth));
+}
+
+TEST(WebSocketServer, WebsocketWorkflow) {
+  MockAuth auth{};
+  communication::websocket::Server websocket_server({"0.0.0.0", 0}, &auth);
+  const auto port = websocket_server.GetEndpoint().port();
+
+  EXPECT_NE(port, 0) << "Port is: " << port;
+  EXPECT_NO_THROW(websocket_server.Start());
+  EXPECT_TRUE(websocket_server.IsRunning());
+
+  EXPECT_NO_THROW(websocket_server.Shutdown());
+  EXPECT_FALSE(websocket_server.IsRunning());
+
+  EXPECT_NO_THROW(websocket_server.AwaitShutdown());
+  EXPECT_FALSE(websocket_server.IsRunning());
+}
+
+TEST(WebSocketServer, WebsocketLoggingSink) {
+  MockAuth auth{};
+  communication::websocket::Server websocket_server({"0.0.0.0", 0}, &auth);
+  EXPECT_NO_THROW(websocket_server.GetLoggingSink());
```

Comment:
Maybe try to write something to the log and read on the other side.

Reply:
I've actually discovered a bug here, thanks